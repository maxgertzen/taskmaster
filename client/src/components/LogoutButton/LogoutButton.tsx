import { useAuth0 } from '@auth0/auth0-react';
import { FC } from 'react';

import { SpriteIcon } from '../SpriteIcon/SpriteIcon';

import { StyledLogoutButton } from './LogoutButton.styled';

export const LogoutButton: FC = () => {
  const { logout } = useAuth0();

  return (
    <StyledLogoutButton
      onClick={() =>
        logout({
          logoutParams: { returnTo: `${window.location.origin}/signin` },
        })
      }
    >
      <SpriteIcon name='logout' />
      Logout
    </StyledLogoutButton>
  );
};
