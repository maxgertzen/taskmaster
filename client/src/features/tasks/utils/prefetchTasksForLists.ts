import { QueryClient } from '@tanstack/react-query';

import { QUERY_KEYS } from '@/shared/api/query-keys';
import { STALE_TIME } from '@/shared/constants/staleTime';

import { fetchTasks } from '../api';

export const prefetchTasksForLists = (
  listIds: string[],
  queryClient: QueryClient,
  token: string | null
) => {
  if (!token || !listIds.length) return;

  listIds.forEach((listId) => {
    queryClient.prefetchQuery({
      queryKey: QUERY_KEYS.tasks({ listId }),
      queryFn: fetchTasks(token),
      staleTime: STALE_TIME,
    });
  });
};
