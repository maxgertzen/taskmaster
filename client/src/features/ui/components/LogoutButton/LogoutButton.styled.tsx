import styled from '@emotion/styled';

export const StyledLogoutButton = styled.button(({ theme }) => ({
  display: 'flex',
  alignItems: 'center',
  color: theme.colors.text,
  gap: theme.spacing(1),
  ':hover': {
    filter: 'invert(1)',
  },
}));
