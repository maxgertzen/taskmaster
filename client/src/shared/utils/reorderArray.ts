export const reorderArray = <T>(
  array: T[],
  fromIndex: number,
  toIndex: number
): T[] => {
  const reordered = [...array];
  const [movedItem] = reordered.splice(fromIndex, 1);
  reordered.splice(toIndex, 0, movedItem);
  return reordered;
};
