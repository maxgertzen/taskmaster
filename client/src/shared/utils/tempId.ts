export const generateTempId = (prefix: string = 'temp') =>
  `${prefix}-${Date.now()}`;
