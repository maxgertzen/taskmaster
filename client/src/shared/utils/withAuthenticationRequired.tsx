import { withAuthenticationRequired as realWithAuthenticationRequired } from '@auth0/auth0-react';

import { IS_AUTH0_DISABLED } from '@/constants';
import { mockWithAuthenticationRequired } from '@/mocks';

const isMock = IS_AUTH0_DISABLED === 'true';

export const withAuthenticationRequired = isMock
  ? mockWithAuthenticationRequired
  : realWithAuthenticationRequired;
