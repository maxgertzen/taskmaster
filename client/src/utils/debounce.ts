export function debounce<T extends (...args: Parameters<T>) => ReturnType<T>>(
  func: T,
  delay: number
): ((...args: Parameters<T>) => void) & { cancel: () => void } {
  let timeoutId: ReturnType<typeof setTimeout> | null = null;

  const debouncedFunction = (...args: Parameters<T>) => {
    if (timeoutId) clearTimeout(timeoutId);
    timeoutId = setTimeout(() => func(...args), delay);
  };

  debouncedFunction.cancel = () => {
    if (timeoutId) clearTimeout(timeoutId);
  };

  return debouncedFunction;
}
