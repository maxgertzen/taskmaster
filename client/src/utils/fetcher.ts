export type FetcherConfig<K> = {
  url: string;
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE';
  body?: K;
  timeout?: number;
  headers?: HeadersInit;
  urlSearchParams?: string;
};

export const fetcher = async <T, K = undefined>(
  config: FetcherConfig<K>
): Promise<T> => {
  try {
    const {
      method = 'GET',
      body = {},
      headers = {},
      timeout = 5000,
      urlSearchParams,
    } = config;
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeout);

    const options: RequestInit = {
      method,
      headers: {
        'Content-Type': 'application/json',
        ...headers,
      },
      ...(method !== 'GET' && { body: JSON.stringify(body) }),
      signal: controller.signal,
    };

    if (urlSearchParams) {
      config.url += `?${urlSearchParams}`;
    }

    const response = await fetch(config.url, options);
    clearTimeout(timeoutId);

    if (!response.ok) {
      throw new Error(`Error: ${response.status} ${response.statusText}`);
    }

    return await response.json();
  } catch (error) {
    console.error('Failed to fetch data:', error);
    throw error;
  }
};
