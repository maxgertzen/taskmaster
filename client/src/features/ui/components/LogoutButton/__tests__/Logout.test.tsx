import { describe, it, expect, vi, beforeEach } from 'vitest';

import { render, screen } from '@/tests/utils';

import { LogoutButton } from '../LogoutButton';

const mockLogout = vi.fn();
vi.mock('@auth0/auth0-react', () => ({
  useAuth0: () => ({
    logout: mockLogout,
  }),
}));

describe('LogoutButton', () => {
  beforeEach(() => {
    vi.clearAllMocks();
    // Mock window.location.origin
    Object.defineProperty(window, 'location', {
      value: { origin: 'http://localhost:3000' },
      writable: true,
    });
  });

  it.each([
    ['light', undefined],
    ['dark', 'dark'],
  ])('renders correctly in %s theme', (_, theme) => {
    const { container } = render(<LogoutButton />, {
      theme: theme as 'light' | 'dark',
    });
    expect(container).toMatchSnapshot();
  });

  describe('User Interactions', () => {
    it('calls logout with correct return URL when clicked', async () => {
      const { user } = render(<LogoutButton />);

      await user.click(screen.getByText('Logout'));

      expect(mockLogout).toHaveBeenCalledWith({
        logoutParams: {
          returnTo: 'http://localhost:3000/signin',
        },
      });
    });
  });

  describe('Rendering', () => {
    it('displays logout text and icon', () => {
      render(<LogoutButton />);

      expect(screen.getByText('Logout')).toBeInTheDocument();
      expect(screen.getByTestId('logout-icon')).toBeInTheDocument();
    });
  });
});
