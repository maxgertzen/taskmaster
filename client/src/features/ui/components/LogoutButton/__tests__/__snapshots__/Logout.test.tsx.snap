// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LogoutButton > renders correctly in dark theme 1`] = `
<div>
  <button
    class="css-3ersaf"
  >
    <div
      class="css-4dcg29"
      data-testid="icon-logout"
      role="button"
      size="3"
    >
      <svg
        data-testid="logout-icon"
      />
    </div>
    Logout
  </button>
</div>
`;

exports[`LogoutButton > renders correctly in light theme 1`] = `
<div>
  <button
    class="css-h00ms8"
  >
    <div
      class="css-4dcg29"
      data-testid="icon-logout"
      role="button"
      size="3"
    >
      <svg
        data-testid="logout-icon"
      />
    </div>
    Logout
  </button>
</div>
`;
